fuzzers/symcc_aflplusplus/fuzzer.py

```diff
@@ -0,0 +1,131 @@
+        # twice in the same directory without this.
+        aflplusplus_fuzzer.build()
+
+    print("Step 2: Completed AFL build")
+    # Copy over AFL artifacts needed by SymCC.
+    shutil.copy("/afl/afl-fuzz", build_directory)
+    shutil.copy("/afl/afl-showmap", build_directory)
+
+    # Build the SymCC-instrumented target.
+    print("Step 3: Building the benchmark with SymCC")
+    symcc_build_dir = get_symcc_build_dir(os.environ['OUT'])
+    os.mkdir(symcc_build_dir)
+
+    # Set flags to ensure compilation with SymCC.
+    new_env['CC'] = "/symcc/build/symcc"
+    new_env['CXX'] = "/symcc/build/sym++"
+    new_env['CXXFLAGS'] = new_env['CXXFLAGS'].replace("-stlib=libc++", "")
+    new_env['FUZZER_LIB'] = '/libfuzzer-harness.o'
+    new_env['OUT'] = symcc_build_dir
+
+    new_env['CXXFLAGS'] += " -fno-sanitize=all "
+    new_env['CFLAGS'] += " -fno-sanitize=all "
+
+    # Setting this environment variable instructs SymCC to use the
+    # libcxx library compiled with SymCC instrumentation.
+    new_env['SYMCC_LIBCXX_PATH'] = "/libcxx_native_build"
+
+    # Instructs SymCC to consider no symbolic inputs at runtime. This is needed
+    # if, for example, some tests are run during compilation of the benchmark.
+    new_env['SYMCC_NO_SYMBOLIC_INPUT'] = "1"
+
+    # Build benchmark.
+    utils.build_benchmark(env=new_env)
+
+    # Copy over symcc artifacts and symbolic libc++.
+    shutil.copy(
+        "/symcc/build//SymRuntime-prefix/src/SymRuntime-build/libSymRuntime.so",
+        symcc_build_dir)
+    shutil.copy("/usr/lib/libz3.so", os.path.join(symcc_build_dir, "libz3.so"))
+    shutil.copy("/libcxx_native_build/lib/libc++.so.1", symcc_build_dir)
+    shutil.copy("/libcxx_native_build/lib/libc++abi.so.1", symcc_build_dir)
+    shutil.copy("/rust/bin/symcc_fuzzing_helper", symcc_build_dir)
+
+
+def launch_afl_thread(input_corpus, output_corpus, target_binary,
+                      additional_flags):
+    """ Simple wrapper for running AFL. """
+    afl_thread = threading.Thread(target=afl_fuzzer.run_afl_fuzz,
+                                  args=(input_corpus, output_corpus,
+                                        target_binary, additional_flags))
+    afl_thread.start()
+    return afl_thread
+
+
+def fuzz(input_corpus, output_corpus, target_binary):
+    """
+    Launches a master and a secondary instance of AFL, as well as
+    the symcc helper.
+    """
+    target_binary_dir = os.path.dirname(target_binary)
+    symcc_workdir = get_symcc_build_dir(target_binary_dir)
+    target_binary_name = os.path.basename(target_binary)
+    symcc_target_binary = os.path.join(symcc_workdir, target_binary_name)
+
+    # Start a master and secondary instance of AFL.
+    # We need both because of the way SymCC works.
+    print('[run_fuzzer] Running AFL for SymCC')
+    aflplusplus_fuzzer.prepare_fuzz_environment(input_corpus)
+    launch_afl_thread(input_corpus, output_corpus, target_binary,
+                      ["-M", "afl-master"])
```

Comment:
-M enables options we dont want on a benchmark run. rather use `-S afl`.
also please add `os.environ['AFL_DISABLE_TRIM'] = "1"`

why are there two launch_afl_thread calls?

Reply:
Thanks, the two `launch_afl_thread` calls is because I follow the documentation from symcc which asks for two AFL instances (one with -M and one with -S): https://github.com/eurecom-s3/symcc/blob/9b20609adab02279c181010c8b1e61a9a9acac62/docs/Fuzzing.txt#L91

Reply:
why would there any value in this? contrary, this makes the comparison even less fair :)
you compare a single run fuzzer with a double run fuzzer + symcc.

already afl + symcc is unfair vs afl.
you would need to compare afl + afl vs afl + symcc.

I did that in an Eclipser run to see if it was really that good, but it turned out that afl++ + afl++ is better than afl++ + Eclipser ...

Reply:
(the instances only have one CPU assigned so that is the right way to test IMHO)

Reply:
If there is a single CPU dedicated then that should make it fair? Interesting with the Eclipser experiment - but again, if there is a single CPU dedicated then it shouldn't matter if you run N instances vs 1? - in fact it should be a disadvantage for the ones with multiple analysis tools assuming more context switches?

Reply:
it makes a difference as a single fuzzer does not 100% utilize the CPU. the more fuzzers you run, the close to the 100% you get. e.g. if you have 10 core * 2 thread = 20 thread machine, you will have peak exec/c at ~25 fuzzing threads in my experience. (with afl++ where we pin to CPUs, with libfuzzer that might be even more because they are less effective on their own)

Reply:
but to make a fairer comparison - add a fuzzer instance that has just the two afl and no symcc - then this is something to compare against (although the 2x afl + symcc should still have an advantage simply because it has a 3rd fuzzer doing things)

Reply:
> but to make a fairer comparison - add a fuzzer instance that has just the two afl and no symcc - then this is something to compare against (although the 2x afl + symcc should still have an advantage simply because it has a 3rd fuzzer doing things)

Yeah I agree - I will add this too so we can that comparison as well. Will also switch to using a single AFL although I will have to check it doesn't do anything unexpected to symcc.